### PR TITLE
SPUI-153 > Add requests signing

### PR DESCRIPTION
**Pull request by VovaMedvediev**

### CODE REVIEW

File: lib/shared/helpers/refresh_token.dart [hunk 1]
```diff
@@ -27,11 +28,22 @@ Future<RefreshTokenStatus> refreshToken(Reader read) async {
...
   final storageService = read(localStorageServicePod);
+  final rsaService = read(rsaServicePod);
+  final serverTime = read(serverTimeSpod);
 
   try {
+    final privateKey = await storageService.getString(privateKeyKey);
+    final refreshToken = authModel.refreshToken;
+    final serverTimeString = serverTime.data?.value.now.toString();
+    final tokenDateTimeSignatureBase64 = await rsaService.sign(
+      refreshToken + (serverTimeString ?? ''),
```
[the remark this answers is not in the document]
Comment (eli1stark) [reply lib/shared/helpers/refresh_token.dart:39]: Rethink this flow please

File: pubspec.yaml [hunk 2]
```diff
@@ -55,6 +55,9 @@ dependencies:
   # NOTIFICATIONS
   flutter_local_notifications: ^6.0.0
 
+  # RSA signing
+  fast_rsa: ^1.3.6
```
Comment (eli1stark) [inline pubspec.yaml:59]: Move to "OTHER" section:
`fast_rsa: ^1.3.6 # RSA signing`

File: pubspec.yaml [hunk 3]
```diff
@@ -81,6 +84,10 @@ dev_dependencies:
 flutter:
   generate: true
   uses-material-design: true
+  assets:
+  - packages/fast_rsa/web/assets/worker.js
```
Comment (eli1stark) [inline pubspec.yaml:88]: Add comment: `// [Start] fast_rsa  -> ... <- fast_rsa [End]`

**Review by eli1stark — CHANGES_REQUESTED** [review]

File: pubspec.yaml (hunk 2 above)
Comment (eli1stark) [reply pubspec.yaml:59]: And add the latest version (2.0.1) with support of null-safety

File: lib/service/services/authentication/model/authenticate/login_request_model.dart [hunk 4]
```diff
@@ -8,7 +8,7 @@ part 'login_request_model.g.dart';
 class LoginRequestModel with _$LoginRequestModel {
   const factory LoginRequestModel({
     String? captcha,
-    String? publicKeyPem,
+    required String publicKeyPem,
```
[the remark this answers is not in the document]
Comment (VovaMedvediev) [reply lib/service/services/authentication/model/authenticate/login_request_model.dart:11]: done

File: lib/shared/helpers/refresh_token.dart [hunk 5]
```diff
@@ -27,11 +28,22 @@ Future<RefreshTokenStatus> refreshToken(Reader read) async {
   final authModelNotifier = read(authModelNotipod.notifier);
   final authService = read(authServicePod);
   final storageService = read(localStorageServicePod);
+  final rsaService = read(rsaServicePod);
+  final serverTime = read(serverTimeSpod);
 
   try {
+    final privateKey = await storageService.getString(privateKeyKey);
+    final refreshToken = authModel.refreshToken;
+    final serverTimeString = serverTime.data?.value.now.toString();
```
[the remark this answers is not in the document]
Comment (VovaMedvediev) [reply lib/shared/helpers/refresh_token.dart:37]: done

File: lib/shared/services/rsa_service.dart [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import 'package:fast_rsa/model/bridge.pbenum.dart';
+import 'package:fast_rsa/rsa.dart';
+import 'local_storage_service.dart';
+
+class RsaService {
+  static const int _bits = 2048;
```
[the remark this answers is not in the document]
Comment (VovaMedvediev) [reply lib/shared/services/rsa_service.dart:6]: done

File: lib/shared/services/local_storage_service.dart [hunk 7]
```diff
@@ -3,6 +3,7 @@ import 'package:shared_preferences/shared_preferences.dart';
 // KEYS
 const refreshTokenKey = 'refreshTokenKey';
```
[the remark this answers is not in the document]
Comment (VovaMedvediev) [reply lib/shared/services/local_storage_service.dart:4]: done

File: lib/shared/dio/helpers/add_signing.dart [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        if (options.method == postRequest) {
+          final requestBody = options.data;
+
+          final rsaService = read(rsaServicePod);
+          final storageService = read(localStorageServicePod);
+
+          final privateKey = await storageService.getString(privateKeyKey);
+          final signature = await rsaService.sign(
+            jsonEncode(requestBody),
+            privateKey ?? '',
```
[the remark this answers is not in the document]
Comment (VovaMedvediev) [reply lib/shared/dio/helpers/add_signing.dart:25]: done

File: lib/shared/helpers/refresh_token.dart [hunk 9]
```diff
@@ -27,11 +28,22 @@ Future<RefreshTokenStatus> refreshToken(Reader read) async {
...
+    final tokenDateTimeSignatureBase64 = await rsaService.sign(
+      refreshToken + (serverTimeString ?? ''),
+      privateKey ?? '',
```
[the remark this answers is not in the document]
Comment (VovaMedvediev) [reply lib/shared/helpers/refresh_token.dart:40]: done

File: lib/shared/helpers/refresh_token.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (VovaMedvediev) [reply lib/shared/helpers/refresh_token.dart:39]: added api for getting server time via rest

File: pubspec.yaml (hunk 2 above)
Comment (VovaMedvediev) [reply pubspec.yaml:59]: done (3.0.0)

File: pubspec.yaml (hunk 3 above)
Comment (VovaMedvediev) [reply pubspec.yaml:88]: done

**Pull request merged**